modules/sdk/src/main/scala/org/tessellation/sdk/domain/healthcheck/consensus/HealthCheckConsensusRound.scala

```diff
@@ -0,0 +1,122 @@
+package org.tessellation.sdk.domain.healthcheck.consensus
+
+import cats.Applicative
+import cats.effect.{Clock, Ref, Spawn}
+import cats.syntax.contravariantSemigroupal._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.sdk.domain.gossip.Gossip
+import org.tessellation.sdk.domain.healthcheck.consensus.types._
+
+class HealthCheckConsensusRound[F[_]: Spawn: Clock, A <: HealthCheckStatus, B <: ConsensusHealthStatus[A]](
+  key: HealthCheckKey,
+  roundId: HealthCheckRoundId,
+  driver: HealthCheckConsensusDriver[A, B],
+  startedAt: Long,
+  peers: Ref[F, Set[PeerId]],
+  roundIds: Ref[F, Set[HealthCheckRoundId]],
+  proposals: Ref[F, Map[PeerId, B]],
+  gossip: Gossip[F],
+  selfId: PeerId
+) {
+
+  def start: F[Unit] =
+    Spawn[F].start {
+      sendProposal
+    }.void
+
+  def getPeers: F[Set[PeerId]] = peers.get
+
+  def manageAbsent(absentPeers: Set[PeerId]): F[Unit] =
+    proposals.get
+      .map(_.keySet)
+      .map(absentPeers -- _)
+      .flatMap { toRemove =>
+        peers.modify { p =>
+          (p -- toRemove, ())
+        }
+      }
+
+  def isFinished: F[Boolean] = allProposalsReceived
+
+  def processProposal(proposal: B): F[Unit] =
+    if (proposal.key == key) {
+      proposals.modify { m =>
+        def updated =
+          m.get(proposal.owner)
+            .fold(m + (proposal.owner -> proposal))(_ => m)
+
+        (updated, ())
+      }.flatMap { _ =>
+        roundIds.modify { m =>
+          (m + proposal.roundId, ())
```

Comment:
there are two cases:
1. proposal already exists
2. proposal does not exist

In case 2 - `m + proposal.roundId` changes nothing, roundId should already be there and it's a `Set` (deduplication)
In case 1 - `m + proposal.roundId` should add the round id

It's the same as you did:
```scala
          roundIds.modify { roundIds => //just adding blindly seems better then checking and conditionally modifying
            (roundIds.add(healthProposal.roundId), ())
          }
```

Reply:
I thinks it's not the same as I did in constellation - I'm only adding the `RoundId` there if the proposal was accepted. Any other case results in `Left` and the `RoundId` is not added. "Blindly adding" being better regarded the fact that if we add the `RoundId` which is there already in the set (because some other node sent its proposal with the same `RoundId` before us) that won't damage anything it will deduplicate. But for this implementation someone could blast us with proposals with different `RoundId`s, proposals which won't be processed/added again, but the round ids for these proposals will get added and it brakes data integrity and may lead to non determinism depending on how will we use the round ids.